### PR TITLE
Test typescript types using tsd

### PR DESCRIPTION
**Pull request by hildjj**
This does not yet have full coverage over all of the types in peg.d.ts;
I want to see if this approach seems viable to the typescript experts.

### CODE REVIEW

**Comment by hildjj** [comment]
Note to future self: to generate the package-lock.json file with an older npm (to keep `lockfileVersion: 1`):

```bash
docker run --rm -it -v $PWD:/app --workdir /app node:12.22.1 npm install --package-lock-only
```

**Comment by hildjj** [comment]
Fixes #155

File: test/types/peg.test-d.ts [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+          expectType<string>(event.rule);
+          switch (event.type) {
+            case "rule.enter":
+              expectType<peggy.LocationRange>(event.location);
+              break;
+            case "rule.match":
+              expectType<peggy.LocationRange>(event.location);
+              expectType<any>(event.result);
+              break;
+          }
```
Comment (Mingun) [inline test/types/peg.test-d.ts:49]: `"rule.fail"` also has a location.
```suggestion
          expectType<peggy.LocationRange>(event.location);
          if (event.type === "rule.match") {
            expectType<any>(event.result);
          }
```

File: test/types/peg.test-d.ts [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+        }
+      }
+    })
+  });
+
+  it("takes an output and grammarSource", () => {
+    const p1 = peggy.generate(src, { output: "parser", grammarSource: "src" });
+    expectType<peggy.Parser>(p1);
+
+    const p2 = peggy.generate(src, { output: "source" });
```
Comment (Mingun) [inline test/types/peg.test-d.ts:59]: Why second test doesn't take a `grammarSource`?
```suggestion
    const p2 = peggy.generate(src, { output: "source", grammarSource: "src" });
  });
```

File: test/types/peg.test-d.ts [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+    expectType<string>(p2);
+  });
+
+  it("creates an AST", () => {
+    const grammar = peggy.parser.parse(src);
+    expectType<peggy.ast.Grammar>(grammar);
+
+    const visited: { [key: string]: number; } = {};
+    function add(typ:string): void {
```
Comment (Mingun) [inline test/types/peg.test-d.ts:68]:
```suggestion
    function add(typ: string): void {
```

File: test/types/peg.test-d.ts [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+
+    const visit = peggy.compiler.visitor.build({
+      grammar(node) {
+        add(node.type);
+        expectType<peggy.ast.Grammar>(node);
+        expectType<"grammar">(node.type);
+        expectType<peggy.LocationRange>(node.location);
+        expectType<peggy.ast.TopLevelInitializer|undefined>(node.topLevelInitializer);
+        expectType<peggy.ast.Initializer|undefined>(node.initializer);
+        expectType<peggy.ast.Rule[]>(node.rules);
```
Comment (Mingun) [inline test/types/peg.test-d.ts:84]: It is necessary to check types of `node` fields? Hasn't the whole `node` test checked them out yet?

File: test/types/peg.test-d.ts [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+        expectType<peggy.LocationRange>(node.location);
+        expectType<string>(node.code);
+        expectType<peggy.LocationRange>(node.codeLocation);
+      },
+      rule(node) {
+        add(node.type);
+        expectType<peggy.ast.Rule>(node);
+        expectType<"rule">(node.type);
+        expectType<peggy.LocationRange>(node.location);
+        expectType<peggy.LocationRange>(node.nameLocation);
```
Comment (Mingun) [inline test/types/peg.test-d.ts:115]: You forgot to check a `name`, if that even needed:
```suggestion
        expectType<peggy.LocationRange>(node.location);
        expectType<string>(node.name);
        expectType<peggy.LocationRange>(node.nameLocation);
```

File: test/types/peg.test-d.ts [hunk 6]
```diff
@@ -0,0 +1,330 @@
...
+      any(node) {
+        add(node.type);
+        expectType<peggy.ast.Any>(node);
+        expectType<"any">(node.type);
+        expectType<peggy.LocationRange>(node.location);
+      }
+    });
+
+    // @dts-jest:pass:snap
+    visit(grammar);
```
Comment (Mingun) [inline test/types/peg.test-d.ts:285]: If I understand correctly, that construction mean to run `visit(grammar)` and assert types with snapshot that was taken on the first run. However that snapshot should be committed and it seems it is not committed

**Review by Mingun — APPROVED** [review]
It seems good, but I'm prefer to split one commit into several. In the current state it changes too much weakly connected things. I suggest to split it into:
- update dependencies
- change format of `package-lock.json`
- switch jest from command-line switches to the config file
- fix `SyntaxError` inheritance
- add `tsd` and the stuff
- update minified files (you can include this step into commit that changed them)

in any order which would build on each step

File: test/types/peg.test-d.ts (hunk 2 above)
Comment (hildjj) [reply test/types/peg.test-d.ts:59]: Willfix, with something not a string.

File: test/types/peg.test-d.ts (hunk 4 above)
Comment (hildjj) [reply test/types/peg.test-d.ts:84]: What I'm trying to achieve is that we'll notice if we break backward-compatibility in the future.

File: test/types/peg.test-d.ts (hunk 6 above)
Comment (hildjj) [reply test/types/peg.test-d.ts:285]: This is a mistake, left over from a previous attempt to use dts-jest.  Will remove.

File: test/types/peg.test-d.ts (hunk 4 above)
Comment (Mingun) [reply test/types/peg.test-d.ts:84]: Makes sense

**Comment by hildjj** [comment]
> @Mingun I think i fixed everything, but I'm interested in your opinion about being able to detect backward-compat breaks.

nm, just saw your comment.  ok to merge?

**Comment by Mingun** [comment]
LGTM. One minor note: is it necessary to generate declaration files? Automatically-generated files seems very useless because almost all types inferred as `any`. Besides, we already has our own file.

**Comment by hildjj** [comment]
I agree that we don't need to generate .d.ts files.  They will have more interesting things in them once we start documenting our interfaces jsdoc-style, but they're never going to generate something close enough to what we currently have to be backward-compatible.

**Pull request merged**